swap/htlc.go

```diff
@@ -200,33 +281,72 @@ func (h *Htlc) GenSuccessWitness(receiverSig []byte,
 	}
 
 	witnessStack := make(wire.TxWitness, 3)
-	witnessStack[0] = append(receiverSig, byte(txscript.SigHashAll))
-	witnessStack[1] = preimage[:]
-	witnessStack[2] = h.Script
+
+	switch h.Version {
+	case HtlcV1:
+		witnessStack[0] = append(receiverSig, byte(txscript.SigHashAll))
+		witnessStack[1] = preimage[:]
+		witnessStack[2] = h.Script
+
+	case HtlcV2:
+		witnessStack[0] = preimage[:]
+		witnessStack[1] = append(receiverSig, byte(txscript.SigHashAll))
+		witnessStack[2] = h.Script
+
+	default:
+		return nil, ErrInvalidScriptVersion
+	}
 
 	return witnessStack, nil
 }
 
 // IsSuccessWitness checks whether the given stack is valid for redeeming the
 // htlc.
 func (h *Htlc) IsSuccessWitness(witness wire.TxWitness) bool {
-	if len(witness) != 3 {
+	if len(witness) != 3 && len(witness) != 4 {
 		return false
 	}
 
-	isTimeoutTx := bytes.Equal([]byte{0}, witness[1])
+	var isTimeoutTx bool
+
+	switch h.Version {
+	case HtlcV1:
+		isTimeoutTx = bytes.Equal([]byte{0}, witness[1])
+
+	case HtlcV2:
+		isTimeoutTx = len(witness) == 4
+
+	default:
+		return false
```

Comment:
If unknown version, return that the witness is a timeout witness?

Reply:
done